### ISSUE TITLE
Incorrect processing of "B3" header

### ISSUE THREAD

**Issue by fbeaufume** [spring-cloud/spring-cloud-sleuth#1452]
TLDR: when a "B3" header is present in a request processed and dispatched by Spring Cloud Gateway to business apps, the ordering of the spans of the distributed trace is incorrect.

Take a microservices app made from 3 parts: a gateway based on Spring Boot and Spring Cloud Gateway, two business services based on Spring Boot. All parts use Spring Cloud Sleuth to get distributed traces (or Spring Cloud Zipkin to publish them to a Zipkin server).

When the app client does not provide the "B3" header, everything works fine.

But when the client (I can reproduce the problem with Postman) or an intermediate infrastructure element (such as **Pivotal Cloud Foundry**) adds the "B3" header such as "1234567812345678-1234567812345678", some business logs print an invalid spanId and the Zipkin UI displays an incorrect spans ordering.

See uploaded snapshot for an exemple (I used short ID for simplicity sake).

![Zipkin issue](https://user-images.githubusercontent.com/25341695/65692336-a7f11900-e072-11e9-8351-e3afe9a0d08d.png)

The problem is experienced with Spring Cloud Greenwich.SR2 and SR3 (I haven't try other versions).

My naive interpretation is that Spring Cloud Gateway forgets that the "B3" header has a special meaning to Spring Cloud Sleuth and forwards it "as is" instead of updating it or (even better I think) removing it.

**Comment by adriancole** [comment]
well captured problem statement. Looking forward to what others think of this as I'm less familiar with this integration.

**Comment by marcingrzejszczak** [comment]
Hmm we instrument the Gateway via a simple filter. We reuse Brave's components to do that: https://github.com/spring-cloud/spring-cloud-sleuth/blob/2.1.x/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceRequestHttpHeadersFilter.java . The only thing that comes to my mind that could be the reason for issues was this issue #1469 . Can you check if the problem persists with the current snapshots (2.1.5 or 2.2.0) ?

**Comment by fbeaufume** [comment]
Ok, I can check.

I tried to find the snapshot jars, but https://repo.spring.io/snapshot/org/springframework/cloud/ seems outdated.

BTW what are the jars I should update ? spring-cloud-sleuth-core and spring-cloud-sleuth-zipkin ?

I tried to clone and build https://github.com/spring-cloud/spring-cloud-sleuth.git but it fails because of some other missing Spring Cloud snapshot dependencies.

What is the easiest way to upgrade my application and do the check ?

**Comment by spencergibb** [comment]
Include the snapshot repo in your build file and then reference Hoxton.BUILD-SNAPSHOT.

The jars are updated (the dates are the dates the folder was created).
See https://repo.spring.io/snapshot/org/springframework/cloud/spring-cloud-sleuth-dependencies/2.2.0.BUILD-SNAPSHOT/

**Comment by fbeaufume** [comment]
Using Spring Cloud Hoxton.BUILD-SNAPSHOT was not enough, I met applicaton start errors. I had to use Spring Boot 2.2.0 instead of the current 2.1.6.

Then I checked and it still fails. I meet the same behavior.

**Comment by marcingrzejszczak** [comment]
Can you show as a sample that replicates the issue?

**Comment by fbeaufume** [comment]
Hi @marcingrzejszczak, I don't have a MWE yet.
I'll try with Spring Cloud Hoxton RC1.

**Comment by fbeaufume** [comment]
Tested with Hoxton RC1, still bugged.

**Comment by marcingrzejszczak** [comment]
Meh, we'll need a sample I guess.

**Comment by fbeaufume** [comment]
Wrote a minimal working example : https://github.com/fbeaufume/b3-issue
